File: UITests/ArchiveE2ETests.swift
import XCTest

/// E2E-тесты для функциональности архива записей
final class ArchiveE2ETests: XCTestCase {
    let app = XCUIApplication()

    override func setUpWithError() throws {
        continueAfterFailure = false
        app.launch()
    }

    // MARK: - Навигация

    /// SC-001: Отображение навигационной панели с ссылками «Главная» и «Архив»
    func testSC001_NavigationBarDisplayed() throws {
        let navHome = app.buttons["nav-home"]
        XCTAssertTrue(navHome.waitForExistence(timeout: 5), "Кнопка «Главная» должна отображаться")

        let navArchive = app.buttons["nav-archive"]
        XCTAssertTrue(navArchive.waitForExistence(timeout: 5), "Кнопка «Архив» должна отображаться")
    }

    /// SC-002: Навигация между главной и архивом
    func testSC002_NavigationBetweenPages() throws {
        let navArchive = app.buttons["nav-archive"]
        XCTAssertTrue(navArchive.waitForExistence(timeout: 5))
        navArchive.tap()

        let archiveTitle = app.staticTexts["archive-title"]
        XCTAssertTrue(archiveTitle.waitForExistence(timeout: 5), "Заголовок «Архив» должен отображаться")

        let navHome = app.buttons["nav-home"]
        XCTAssertTrue(navHome.waitForExistence(timeout: 5))
        navHome.tap()

        let homeTitle = app.staticTexts["home-title"]
        XCTAssertTrue(homeTitle.waitForExistence(timeout: 5), "Заголовок «Notes App» должен отображаться")
    }

    /// SC-003: Главная страница содержит заголовок, ссылку на архив и демо-секцию
    func testSC003_HomePageStructure() throws {
        let homeTitle = app.staticTexts["home-title"]
        XCTAssertTrue(homeTitle.waitForExistence(timeout: 5))

        let archiveLink = app.buttons["home-archive-link"]
        XCTAssertTrue(archiveLink.waitForExistence(timeout: 5), "Ссылка «Перейти в Архив» должна отображаться")

        let demoSection = app.otherElements["demo-section"]
        XCTAssertTrue(demoSection.waitForExistence(timeout: 5), "Демо-секция должна отображаться")

        let archiveButton = app.buttons["archive-button-demo-1"]
        XCTAssertTrue(archiveButton.waitForExistence(timeout: 5), "Кнопка архивации демо-записи должна отображаться")
    }

    /// SC-004: Ссылка «Перейти в Архив» ведёт на страницу архива
    func testSC004_HomeArchiveLinkNavigation() throws {
        let archiveLink = app.buttons["home-archive-link"]
        XCTAssertTrue(archiveLink.waitForExistence(timeout: 5))
        archiveLink.tap()

        let archiveTitle = app.staticTexts["archive-title"]
        XCTAssertTrue(archiveTitle.waitForExistence(timeout: 5), "Должна открыться страница архива")
    }

    // MARK: - UI архива

    /// SC-201: Страница архива — пустое состояние
    func testSC201_ArchiveEmptyState() throws {
        let navArchive = app.buttons["nav-archive"]
        XCTAssertTrue(navArchive.waitForExistence(timeout: 5))
        navArchive.tap()

        let archiveTitle = app.staticTexts["archive-title"]
        XCTAssertTrue(archiveTitle.waitForExistence(timeout: 5))

        let emptyArchive = app.otherElements["empty-archive"]
        XCTAssertTrue(emptyArchive.waitForExistence(timeout: 5), "Пустое состояние архива должно отображаться")
    }

    /// SC-202: Архивация записи через демо-кнопку
    func testSC202_ArchiveNoteViaDemoButton() throws {
        let archiveButton = app.buttons["archive-button-demo-1"]
        XCTAssertTrue(archiveButton.waitForExistence(timeout: 5))

        archiveButton.tap()

        // Переход на страницу архива
        let navArchive = app.buttons["nav-archive"]
        XCTAssertTrue(navArchive.waitForExistence(timeout: 5))
        navArchive.tap()

        let archiveCard = app.otherElements["archive-card-demo-1"]
        XCTAssertTrue(archiveCard.waitForExistence(timeout: 5), "Карточка архивной записи должна появиться")
    }

    /// SC-203: Восстановление записи из архива
    func testSC203_RestoreNoteFromArchive() throws {
        // Сначала архивируем запись
        let archiveButton = app.buttons["archive-button-demo-1"]
        XCTAssertTrue(archiveButton.waitForExistence(timeout: 5))
        archiveButton.tap()

        // Переходим в архив
        let navArchive = app.buttons["nav-archive"]
        XCTAssertTrue(navArchive.waitForExistence(timeout: 5))
        navArchive.tap()

        let archiveCard = app.otherElements["archive-card-demo-1"]
        XCTAssertTrue(archiveCard.waitForExistence(timeout: 5))

        // Нажимаем кнопку восстановления
        let restoreButton = app.buttons["restore-note-demo-1"]
        XCTAssertTrue(restoreButton.waitForExistence(timeout: 5))
        restoreButton.tap()

        // Проверяем пустое состояние
        let emptyArchive = app.otherElements["empty-archive"]
        XCTAssertTrue(emptyArchive.waitForExistence(timeout: 5), "Архив должен стать пустым после восстановления")
    }

    /// SC-204: Карточка архивной записи — отображение данных
    func testSC204_ArchiveCardDisplaysData() throws {
        // Архивируем демо-запись
        let archiveButton = app.buttons["archive-button-demo-1"]
        XCTAssertTrue(archiveButton.waitForExistence(timeout: 5))
        archiveButton.tap()

        // Переходим в архив
        let navArchive = app.buttons["nav-archive"]
        XCTAssertTrue(navArchive.waitForExistence(timeout: 5))
        navArchive.tap()

        let archiveCard = app.otherElements["archive-card-demo-1"]
        XCTAssertTrue(archiveCard.waitForExistence(timeout: 5))

        // Проверяем содержимое карточки
        XCTAssertTrue(archiveCard.staticTexts["Демо запись"].exists, "Заголовок записи должен отображаться")
        XCTAssertTrue(archiveCard.staticTexts["Содержимое демо записи"].exists, "Содержимое записи должно отображаться")

        let restoreButton = app.buttons["restore-note-demo-1"]
        XCTAssertTrue(restoreButton.waitForExistence(timeout: 5), "Кнопка восстановления должна отображаться")
    }

    /// SC-205: Данные сохраняются после перезапуска приложения
    func testSC205_PersistenceAfterRelaunch() throws {
        // Архивируем запись
        let archiveButton = app.buttons["archive-button-demo-1"]
        XCTAssertTrue(archiveButton.waitForExistence(timeout: 5))
        archiveButton.tap()

        // Перезапускаем приложение
        app.terminate()
        app.launch()

        // Переходим в архив
        let navArchive = app.buttons["nav-archive"]
        XCTAssertTrue(navArchive.waitForExistence(timeout: 5))
        navArchive.tap()

        // Проверяем, что карточка всё ещё отображается
        let archiveCard = app.otherElements["archive-card-demo-1"]
        XCTAssertTrue(archiveCard.waitForExistence(timeout: 5), "Архивная запись должна сохраниться после перезапуска")
    }
}

File: android/src/androidTest/java/com/example/testfullstackapp/ArchiveE2ETests.kt
package com.example.testfullstackapp

import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.assertIsNotDisplayed
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

/**
 * E2E-тесты для функциональности архива записей
 */
@RunWith(AndroidJUnit4::class)
class ArchiveE2ETests {

    @get:Rule
    val composeTestRule = createAndroidComposeRule<MainActivity>()

    // MARK: - Навигация

    /** SC-001: Отображение навигационной панели с ссылками «Главная» и «Архив» */
    @Test
    fun testSC001_NavigationBarDisplayed() {
        composeTestRule.onNodeWithTag("nav-home").assertIsDisplayed()
        composeTestRule.onNodeWithTag("nav-archive").assertIsDisplayed()
    }

    /** SC-002: Навигация между главной и архивом */
    @Test
    fun testSC002_NavigationBetweenPages() {
        // Переход на архив
        composeTestRule.onNodeWithTag("nav-archive").performClick()
        composeTestRule.onNodeWithTag("archive-title").assertIsDisplayed()

        // Возврат на главную
        composeTestRule.onNodeWithTag("nav-home").performClick()
        composeTestRule.onNodeWithTag("home-title").assertIsDisplayed()
    }

    /** SC-003: Главная страница содержит заголовок, ссылку на архив и демо-секцию */
    @Test
    fun testSC003_HomePageStructure() {
        composeTestRule.onNodeWithTag("home-title").assertIsDisplayed()
        composeTestRule.onNodeWithTag("home-archive-link").assertIsDisplayed()
        composeTestRule.onNodeWithTag("demo-section").assertIsDisplayed()
        composeTestRule.onNodeWithTag("archive-button-demo-1").assertIsDisplayed()
    }

    /** SC-004: Ссылка «Перейти в Архив» ведёт на страницу архива */
    @Test
    fun testSC004_HomeArchiveLinkNavigation() {
        composeTestRule.onNodeWithTag("home-archive-link").performClick()
        composeTestRule.onNodeWithTag("archive-title").assertIsDisplayed()
    }

    // MARK: - UI архива

    /** SC-201: Страница архива — пустое состояние */
    @Test
    fun testSC201_ArchiveEmptyState() {
        composeTestRule.onNodeWithTag("nav-archive").performClick()

        composeTestRule.onNodeWithTag("archive-title").assertIsDisplayed()
        composeTestRule.onNodeWithTag("empty-archive").assertIsDisplayed()
        composeTestRule.onNodeWithText("Архив пуст").assertIsDisplayed()
        composeTestRule.onNodeWithText("Архивируйте записи, чтобы они появились здесь").assertIsDisplayed()
    }

    /** SC-202: Архивация записи через демо-кнопку */
    @Test
    fun testSC202_ArchiveNoteViaDemoButton() {
        // Нажимаем кнопку архивации
        composeTestRule.onNodeWithTag("archive-button-demo-1").performClick()

        // Переходим в архив
        composeTestRule.onNodeWithTag("nav-archive").performClick()

        // Проверяем карточку
        composeTestRule.onNodeWithTag("archive-card-demo-1").assertIsDisplayed()
        composeTestRule.onNodeWithText("Демо запись").assertIsDisplayed()
    }

    /** SC-203: Восстановление записи из архива */
    @Test
    fun testSC203_RestoreNoteFromArchive() {
        // Архивируем запись
        composeTestRule.onNodeWithTag("archive-button-demo-1").performClick()

        // Переходим в архив
        composeTestRule.onNodeWithTag("nav-archive").performClick()
        composeTestRule.onNodeWithTag("archive-card-demo-1").assertIsDisplayed()

        // Восстанавливаем
        composeTestRule.onNodeWithTag("restore-note-demo-1").performClick()

        // Проверяем пустое состояние
        composeTestRule.onNodeWithTag("empty-archive").assertIsDisplayed()
    }

    /** SC-204: Карточка архивной записи — отображение данных */
    @Test
    fun testSC204_ArchiveCardDisplaysData() {
        // Архивируем демо-запись
        composeTestRule.onNodeWithTag("archive-button-demo-1").performClick()

        // Переходим в архив
        composeTestRule.onNodeWithTag("nav-archive").performClick()

        // Проверяем содержимое карточки
        composeTestRule.onNodeWithTag("archive-card-demo-1").assertIsDisplayed()
        composeTestRule.onNodeWithText("Демо запись").assertIsDisplayed()
        composeTestRule.onNodeWithText("Содержимое демо записи").assertIsDisplayed()
        composeTestRule.onNodeWithTag("restore-note-demo-1").assertIsDisplayed()
    }

    /** SC-205: Данные сохраняются после перезапуска (пересоздания Activity) */
    @Test
    fun testSC205_PersistenceAfterRecreation() {
        // Архивируем запись
        composeTestRule.onNodeWithTag("archive-button-demo-1").performClick()

        // Пересоздаём Activity для проверки persistence
        composeTestRule.activityRule.scenario.recreate()

        // Переходим в архив
        composeTestRule.onNodeWithTag("nav-archive").performClick()

        // Проверяем, что карточка сохранилась
        composeTestRule.onNodeWithTag("archive-card-demo-1").assertIsDisplayed()
    }
}
